File: dotnet/src/Http.cs
// Minimal blocking HTTP client standing in for the global `fetch`.
//
// Like Node's fetch: no system proxy, redirects followed, gzip/deflate/br
// decoded, User-Agent "skills-cli/<version>" unless a request sets its own.

using System.Net;
using System.Text;
using System.Text.Json.Nodes;

namespace Skills;

internal sealed class HttpResponse(int status, Dictionary<string, string> headers, byte[] body)
{
    public int Status { get; } = status;
    public byte[] Body { get; } = body;
    public bool Ok => Status is >= 200 and < 300;

    public string? Header(string name) => headers.TryGetValue(name.ToLowerInvariant(), out var v) ? v : null;

    public string Text() => DecodeUtf8(Body);

    /// `response.json()`: UTF-8 decode (dropping a BOM) then JSON.parse.
    public JsonNode? Json() => Skills.Json.TryParse(Text(), out var n) ? n : null;

    public bool TryJson(out JsonNode? node) => Skills.Json.TryParse(Text(), out node);

    /// TextDecoder('utf-8'): replacement characters, leading BOM removed.
    public static string DecodeUtf8(byte[] bytes)
    {
        var s = Encoding.UTF8.GetString(bytes);
        return s.StartsWith('﻿') ? s[1..] : s;
    }
}

internal sealed class HttpException(string message) : Exception(message);

internal sealed class HttpTooLargeException(long max) : Exception($"too-large:{max}")
{
    public long Max { get; } = max;
}

internal sealed class HttpRequest(string url)
{
    private static readonly HttpClient Client = new(new SocketsHttpHandler
    {
        AllowAutoRedirect = true,
        MaxAutomaticRedirections = 20,
        AutomaticDecompression = DecompressionMethods.All,
        UseProxy = false,
        PooledConnectionLifetime = TimeSpan.FromMinutes(5),
    })
    {
        Timeout = System.Threading.Timeout.InfiniteTimeSpan,
    };

    private readonly List<(string Key, string Value)> _headers = new();
    private TimeSpan _timeout = TimeSpan.FromSeconds(300);
    private long? _maxBytes;

    public static HttpRequest Get(string url) => new(url);

    public HttpRequest Header(string k, string v)
    {
        _headers.Add((k, v));
        return this;
    }

    public HttpRequest Timeout(TimeSpan t)
    {
        _timeout = t;
        return this;
    }

    public HttpRequest MaxBytes(long n)
    {
        _maxBytes = n;
        return this;
    }

    /// Perform the request. Transport failures throw <see cref="HttpException"/>;
    /// any HTTP status is a response. Oversized successful bodies throw
    /// <see cref="HttpTooLargeException"/>.
    public HttpResponse Send()
    {
        try
        {
            return SendAsync().GetAwaiter().GetResult();
        }
        catch (HttpTooLargeException)
        {
            throw;
        }
        catch (Exception e) when (e is HttpRequestException or TaskCanceledException or OperationCanceledException or IOException or UriFormatException or InvalidOperationException or NotSupportedException)
        {
            throw new HttpException(e is TaskCanceledException or OperationCanceledException ? "The operation was aborted due to timeout" : e.Message);
        }
    }

    private async Task<HttpResponse> SendAsync()
    {
        using var cts = new CancellationTokenSource(_timeout);
        using var req = new HttpRequestMessage(HttpMethod.Get, url);
        var hasUa = false;
        foreach (var (k, v) in _headers)
        {
            if (k.Equals("User-Agent", StringComparison.OrdinalIgnoreCase)) hasUa = true;
            req.Headers.TryAddWithoutValidation(k, v);
        }
        if (!hasUa) req.Headers.TryAddWithoutValidation("User-Agent", $"skills-cli/{Program.Version}");
        using var resp = await Client.SendAsync(req, HttpCompletionOption.ResponseHeadersRead, cts.Token).ConfigureAwait(false);
        var status = (int)resp.StatusCode;
        var headers = new Dictionary<string, string>();
        foreach (var h in resp.Headers) headers[h.Key.ToLowerInvariant()] = string.Join(", ", h.Value);
        foreach (var h in resp.Content.Headers) headers[h.Key.ToLowerInvariant()] = string.Join(", ", h.Value);

        // Error statuses are reported before size limits (the TS download checks
        // `response.ok` first); for size-limited requests their bodies are skipped.
        var ok = status is >= 200 and < 300;
        var max = ok ? _maxBytes : null;
        if (!ok && _maxBytes != null) return new HttpResponse(status, headers, []);
        if (max != null && resp.Content.Headers.ContentLength is { } len && len > max) throw new HttpTooLargeException(max.Value);

        await using var stream = await resp.Content.ReadAsStreamAsync(cts.Token).ConfigureAwait(false);
        using var ms = new MemoryStream();
        var buffer = new byte[81920];
        int read;
        while ((read = await stream.ReadAsync(buffer, cts.Token).ConfigureAwait(false)) > 0)
        {
            ms.Write(buffer, 0, read);
            if (max != null && ms.Length > max) throw new HttpTooLargeException(max.Value);
        }
        return new HttpResponse(status, headers, ms.ToArray());
    }

    /// Send, returning null on transport failure (the common `try { fetch } catch {}`).
    public HttpResponse? TrySend()
    {
        try
        {
            return Send();
        }
        catch (HttpException)
        {
            return null;
        }
        catch (HttpTooLargeException)
        {
            return null;
        }
    }
}

File: rust/src/http.rs
//! Minimal blocking HTTP client standing in for the global `fetch`.

use std::io::Read;
use std::time::Duration;

pub struct Response {
    pub status: u16,
    headers: Vec<(String, String)>,
    pub body: Vec<u8>,
}

impl Response {
    pub fn ok(&self) -> bool {
        (200..300).contains(&self.status)
    }

    pub fn header(&self, name: &str) -> Option<&str> {
        let lower = name.to_lowercase();
        self.headers
            .iter()
            .find(|(k, _)| k.to_lowercase() == lower)
            .map(|(_, v)| v.as_str())
    }

    pub fn text(&self) -> String {
        String::from_utf8_lossy(&self.body).into_owned()
    }

    pub fn json(&self) -> Option<serde_json::Value> {
        // `Response.json()` decodes as UTF-8, which drops a leading BOM.
        let body = self
            .body
            .strip_prefix(b"\xEF\xBB\xBF")
            .unwrap_or(&self.body);
        serde_json::from_slice(body).ok()
    }
}

pub struct Request<'a> {
    url: &'a str,
    headers: Vec<(String, String)>,
    timeout: Duration,
    max_bytes: Option<u64>,
}

/// Error raised when a response body exceeds `max_bytes`.
#[derive(Debug)]
pub struct TooLarge;

pub fn get(url: &str) -> Request<'_> {
    Request {
        url,
        headers: Vec::new(),
        timeout: Duration::from_secs(300),
        max_bytes: None,
    }
}

impl<'a> Request<'a> {
    pub fn header(mut self, k: &str, v: &str) -> Self {
        self.headers.push((k.to_string(), v.to_string()));
        self
    }

    pub fn timeout(mut self, d: Duration) -> Self {
        self.timeout = d;
        self
    }

    pub fn max_bytes(mut self, n: u64) -> Self {
        self.max_bytes = Some(n);
        self
    }

    /// Perform the request. Transport failures are `Err`; any HTTP status is `Ok`.
    pub fn send(self) -> Result<Response, String> {
        let agent = ureq::AgentBuilder::new()
            .timeout(self.timeout)
            .redirects(20)
            .user_agent(concat!("skills-cli/", env!("CARGO_PKG_VERSION")))
            .build();
        let mut req = agent.get(self.url);
        for (k, v) in &self.headers {
            req = req.set(k, v);
        }
        let resp = match req.call() {
            Ok(r) => r,
            Err(ureq::Error::Status(_, r)) => r,
            Err(e) => return Err(e.to_string()),
        };
        let status = resp.status();
        let headers: Vec<(String, String)> = resp
            .headers_names()
            .iter()
            .filter_map(|n| resp.header(n).map(|v| (n.clone(), v.to_string())))
            .collect();
        // Error statuses are reported before size limits (the TS download checks
        // `response.ok` first); for size-limited requests their bodies are skipped.
        let max_bytes = if (200..300).contains(&status) {
            self.max_bytes
        } else {
            None
        };
        if let (Some(max), Some(len)) = (
            max_bytes,
            resp.header("content-length")
                .and_then(|l| l.parse::<u64>().ok()),
        ) {
            if len > max {
                return Err(format!("too-large:{}", max));
            }
        }
        let mut body = Vec::new();
        let reader = resp.into_reader();
        let limit = match (max_bytes, self.max_bytes) {
            (Some(m), _) => m + 1,
            // Size-limited request that failed: the caller only needs the status.
            (None, Some(_)) => 0,
            (None, None) => u64::MAX,
        };
        reader
            .take(limit)
            .read_to_end(&mut body)
            .map_err(|e| e.to_string())?;
        if let Some(max) = max_bytes {
            if body.len() as u64 > max {
                return Err(format!("too-large:{}", max));
            }
        }
        Ok(Response {
            status,
            headers,
            body,
        })
    }
}
